File: generated_code/GeminiPro/Python/060_SplayTree.py
import sys

# To prevent recursion depth errors on large trees, though unlikely with splaying
# sys.setrecursionlimit(2000)

class SplayTree:
    """
    An implementation of a Splay Tree that acts as a dictionary-like set for integers.
    Splay trees are self-balancing binary search trees with the additional property
    that recently accessed elements are quick to access again.

    This implementation includes insert, delete, and search operations, with the
    characteristic splaying operation performed on each access.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag case
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.
        
        Performs the splaying operation on the accessed node if found,
        or on its parent if not found.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                break
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        
        if last_node:
            self._splay(last_node)
        
        return self.root and self.root.key == key

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed
        to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay the node and return
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Insert the new node
        new_node = self._Node(key)
        new_node.parent = parent
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The standard splay tree deletion algorithm is used:
        1. Search for the key, which splays the node (or a nearby one) to the root.
        2. If the key is not at the root, it wasn't in the tree.
        3. If the key is at the root, remove it and join the two resulting subtrees.
           - To join, find the maximum element in the left subtree.
           - Splay this maximum element to the root of the left subtree.
           - Attach the right subtree as the right child of this new root.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Search splays the node or its parent to the root
        self.search(key)

        # If key is not at the root after splaying, it's not in the tree
        if self.root.key != key:
            return

        # Now the node to delete is at the root
        node_to_delete = self.root
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            # Promote the right subtree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # Promote the left subtree
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Join the two subtrees
            left_subtree.parent = None
            
            # Find the max element in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max element to the root of the left subtree
            # by temporarily treating the left subtree as the main tree.
            self.root = left_subtree
            self._splay(max_node)
            
            # Now self.root is the max_node. Attach the original right subtree.
            self.root.right = right_subtree
            right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/060_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 *
 * A splay tree is a self-adjusting binary search tree with the additional
 * property that recently accessed elements are quick to access again. It performs
 * basic operations such as insertion, look-up and removal in O(log n) amortized time.
 *
 * This class provides the following public methods:
 * - {@code insert(int key)}: Adds a key to the set.
 * - {@code delete(int key)}: Removes a key from the set.
 * - {@code search(int key)}: Searches for a key and returns it, or null if not found.
 *
 * The search, insert, and delete operations all perform a "splay" operation, which
 * moves the accessed (or inserted/deleted) node to the root of the tree to optimize
 * future accesses.
 */
public class SplayTree {

    private Node root;

    /**
     * Represents a node in the splay tree.
     * This is a private static inner class as it's only used by SplayTree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods for Rotations ---

    /**
     * Performs a right rotation on the given node `x`.
     * This is a standard tree rotation operation.
     *
     * @param x The node to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rotateRight(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the given node `x`.
     * This is a standard tree rotation operation.
     *
     * @param x The node to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rotateLeft(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splaying operation. It moves the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This implementation uses the top-down splaying approach, which restructures
     * the tree as it searches for the key from the root down.
     *
     * @param key The key to splay to the root.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // Create a dummy node to simplify linking.
        // The left child of the dummy will be the root of the "right tree" (R).
        // The right child of the dummy will be the root of the "left tree" (L).
        Node dummy = new Node(0); // Key doesn't matter
        Node leftTreeMax = dummy;
        Node rightTreeMin = dummy;
        Node current = root;

        while (true) {
            if (key < current.key) {
                if (current.left == null) break;
                // Zig-Zig case (left-left)
                if (key < current.left.key) {
                    current = rotateRight(current);
                    if (current.left == null) break;
                }
                // Add current to the right tree (R)
                rightTreeMin.left = current;
                rightTreeMin = current;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) break;
                // Zig-Zig case (right-right)
                if (key > current.right.key) {
                    current = rotateLeft(current);
                    if (current.right == null) break;
                }
                // Add current to the left tree (L)
                leftTreeMax.right = current;
                leftTreeMax = current;
                current = current.right;
            } else { // key == current.key
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = current.left;
        rightTreeMin.left = current.right;
        current.left = dummy.right;
        current.right = dummy.left;
        root = current;
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the splay tree.
     * If the key already exists in the set, this operation has no effect
     * besides splaying the existing node to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        splay(key);

        // If key is already present, we are done (it's now the root).
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is found and deleted, the parent of the deleted node (or the
     * predecessor, if the deleted node was the root) is splayed to the root.
     * If the key is not found, the tree is splayed on the last accessed node.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        splay(key);

        // If the key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // Now the node to be deleted is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
        } else {
            // Make the left subtree the new tree.
            root = leftSubtree;
            // Splay for the (now non-existent) key. This brings the largest element
            // in the left subtree (the predecessor of the deleted key) to its root.
            splay(key);
            
            // The new root (predecessor) will have a null right child.
            // Attach the original right subtree there.
            root.right = rightSubtree;
        }
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to the root.
     * If not found, the last accessed node in the search path is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        // After splaying, the root will be the key if it exists.
        if (root.key == key) {
            return key;
        }
        return null;
    }
}